### PR TITLE
Clang-format and include-what-you-use

### PR DESCRIPTION
**Pull request by sethrj**
Some formatting has drifted over time, and we've used various versions of clang-format: this reformats everything, after adding IWYU (below) and enabling include reordering.

I added a configuration to run IWYU but it's not 100% successful:
- Running on some test files caused it to segfault
- It incorrectly reports transitive includes for `detail/RangeImpl.hh`
- It thinks `move` belongs to `type_traits`
- It doesn't format suggested includes as relative directories
- It ignores requirements from disabled preprocessor blocks
- It doesn't have the right relocations for some headers with apple-clang@13 -- suggesting that I include e.g. `<__map>` or whatnot
- The cmake integration won't allow us to automatically apply the changes.

### CODE REVIEW

**Review by drbenmorgan — APPROVED** [review]
Eeee... that's a lot of files 😅 ! Looks fine from a quick sample though.

I'm not sure the issues with IWYU are easily resolvable, if at all in some cases (e.g. `#include`s dependent on current configuration `#define`s). As far as getting the build to apply the fixes, could the [`fix_includes.py`](https://github.com/include-what-you-use/include-what-you-use#applying-fixes) be used through a `add_custom_command/target`? Looks awkward though as the build output is needed as input...

Alternately, could a CI job consider any IWYU output as "failure" (within some regex)? Could give lots of false positives judging by the issues seen so far though.

Approving, but I leave at your discretion to merge/update.

**Comment by sethrj** [comment]
Thanks @drbenmorgan , just returning the favor 😉 

 I bet that getting a CI job to "90%" working wouldn't be too bad, but getting it to 100% seems like a nightmare -- so it might be good to just periodically run this and see if any egregious changes pop up. I thought that now might be an opportune time since this week I think I will start the source code reorganization discussed in #355 .

I'm going to wait until  #374 gets merged for this, since that PR touched a fair number of header blocks.

**Pull request merged**
